rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -0,0 +1,452 @@
+    z_loan(session_),
+    z_keyexpr(liveliness_str_.c_str()),
+    z_move(callback),
+    &sub_options);
+  zc_liveliness_subscriber_options_drop(z_move(sub_options));
+  auto undeclare_z_sub = rcpputils::make_scope_exit(
+    [this]() {
+      z_undeclare_subscriber(z_move(this->graph_subscriber_));
+    });
+  if (!z_check(graph_subscriber_)) {
+    RMW_SET_ERROR_MSG("unable to create zenoh subscription");
+    return RMW_RET_ERROR;
+  }
+
+  undeclare_z_sub.cancel();
+  is_initialized_ = true;
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+rmw_ret_t rmw_context_impl_s::Data::shutdown()
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (is_shutdown_) {
+    return RMW_RET_OK;
+  }
+
+  z_undeclare_subscriber(z_move(graph_subscriber_));
+  if (shm_manager_.has_value()) {
+    z_drop(z_move(shm_manager_.value()));
+  }
+  // Close the zenoh session
+  if (z_close(z_move(session_)) < 0) {
+    RMW_SET_ERROR_MSG("Error while closing zenoh session");
+    return RMW_RET_ERROR;
+  }
+  is_shutdown_ = true;
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+rmw_context_impl_s::Data::~Data()
+{
+  RMW_TRY_DESTRUCTOR(
+    static_cast<rmw_zenoh_cpp::GuardCondition *>(
+      graph_guard_condition_->data)->~GuardCondition(),
+    rmw_zenoh_cpp::GuardCondition, );
+  if (rcutils_allocator_is_valid(allocator_)) {
+    allocator_->deallocate(graph_guard_condition_->data, allocator_->state);
+    allocator_->deallocate(graph_guard_condition_, allocator_->state);
+    graph_guard_condition_ = nullptr;
+  }
+
+  auto ret = this->shutdown();
+  static_cast<void>(ret);
+}
+
+///=============================================================================
+rmw_context_impl_s::rmw_context_impl_s(
+  const rcutils_allocator_t * allocator,
+  const std::size_t domain_id,
+  const std::string & enclave,
+  z_owned_session_t session,
+  std::optional<zc_owned_shm_manager_t> shm_manager,
+  rmw_guard_condition_t * graph_guard_condition)
+{
+  data_ = std::make_shared<Data>(
+    allocator,
+    std::move(domain_id),
+    std::move(enclave),
+    std::move(session),
+    std::move(shm_manager),
+    graph_guard_condition);
+
+  // TODO(Yadunund): Consider switching to a make() pattern to avoid throwing
+  // errors.
+  auto ret = data_->subscribe();
+  if (ret != RMW_RET_OK) {
+    throw std::runtime_error("Unable to subscribe to ROS Graph updates.");
+  }
+}
+
+///=============================================================================
+std::string rmw_context_impl_s::enclave() const
+{
+  std::lock_guard<std::mutex> lock(data_->mutex_);
+  return data_->enclave_;
+}
+
+///=============================================================================
+z_session_t rmw_context_impl_s::session() const
+{
+  std::lock_guard<std::mutex> lock(data_->mutex_);
+  return z_loan(data_->session_);
```

Comment:
This lock isn't doing too much.  The problem is that as soon as the call to `rmw_context_impl_s::session()` returns, the lock is dropped.  At that point, the caller has a loan to the session, but if some other thread changes the session, the caller now has a dangling loan.  We'll need to do something more clever with sessions here in order to make it thread safe.

This same comment applies to the `shm_manager()` and `graph_guard_condition()` methods below (it doesn't apply to `enclave()` above because that is returning a copy).

Reply:
Yes I'm aware of this issue and this is a temporary workaround (I left a similar comment in the function that returns the `std::optional<shm_manager>`).

My goal in subsequent PRs is to add functions like `rmw_context_impl_s::add_publisher()`, `rmw_context_impl_s::publish()` etc which will create/handle the publisher/subscribption/client/service instance inside `rmw_context_impl_s` directly without returning anything. This way we will lock the `session` -> create the instance -> add it to an internal map like in #259.

Reply:
ie this comment https://github.com/ros2/rmw_zenoh/blob/a9b34e2ede699450cc5dd968e87f172827cdbb28/rmw_zenoh_cpp/src/detail/rmw_context_impl_s.hpp#L54-L58